### PR TITLE
Add front-end testing guide

### PR DESCRIPTION
**Pull request by contolini**
This adds a brief testing guide. It links to OAH's repo a lot because they have decent full-stack code coverage. The browser testing section could use some more examples.

Review: @imuchnik

### CODE REVIEW

**Comment by jimmynotjim** [comment]
Should we consider a single framework to reduce differences between projects or is that too stiff of a position to take? I ask because we're working on adding JS testing to FlapJack and are planning on Mocha because OaH and eRegs are using it.

File: testing.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+### Test frameworks
+- [Mocha](http://mochajs.org) is a slick, simple testing framework for Node and browser JavaScript with asynchronous support and a [slew of reporters](http://mochajs.org/#reporters), including [TAP](http://en.wikipedia.org/wiki/Test_Anything_Protocol) output. See the [node-mocha example](examples/node-mocha) for usage.
```
Comment (ascott1) [inline testing.md]: There's a link to a node-mocha example, but I don't see that file.

Comment (contolini) [inline testing.md]: Good catch, that line was taken from the 18f guide linked to at the bottom of the list.

**Comment by contolini** [comment]
@jimmynotjim It personally doesn't bother me if different projects using different test frameworks. Each framework has its pros and cons, each developer has personal preferences, and each project has different needs.

Just be a responsible coder. Take the time to read the docs of the framework you choose. Avoid immature and/or poorly-maintained frameworks. [Embrace boredom](http://mcfunley.com/choose-boring-technology). Add documentation if you're doing something unusual. Consult peers if you need help or advice. And whatever you do: don't use coffeescript.

**Comment by ascott1** [comment]
> And whatever you do: don't use coffeescript.

**Comment by sebworks** [comment]
lol, don't diss CoffeeScript. Much of ES6 has been influenced by it.  https://brendaneich.com/tag/javascript-ecmascript-harmony-coffeescript/

File: testing.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+### Test frameworks
+- [Mocha](http://mochajs.org) is a slick, simple testing framework for Node and browser JavaScript with asynchronous support and a [slew of reporters](http://mochajs.org/#reporters), including [TAP](http://en.wikipedia.org/wiki/Test_Anything_Protocol) output.
+- [Jasmine](https://github.com/jasmine/jasmine) shares basic semantics with [Mocha](http://mochajs.org), but also includes lots of nice assertion helpers, object [spies](http://jasmine.github.io/2.2/introduction.html#section-Spies), and async support.
+- [Tape](https://www.npmjs.com/package/tape) is a [TAP](http://en.wikipedia.org/wiki/Test_Anything_Protocol)-producing test harness for Node with great asynchronous support. Check out the [Testling tape guide](https://ci.testling.com/guide/tape).
+
+Check out 18f's excellent [testing cookbook](http://18f.github.io/testing-cookbook) for more best practices.
```
Comment (anselmbradford) [inline testing.md]: Hm, I get a 404 for that URL for the testing-cookbook. The [repo](https://github.com/18F/testing-cookbook) makes it seem like it should  be right though.

File: testing.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+A unit test verifies that a single discrete piece of your codebase works as expected. That "piece" could be a JavaScript function, a CommonJS/AMD module or some other unit of functionality. It's vaguely defined but don't let that scare you. Start by choosing one of the test frameworks below and writing tests for individual JavaScript methods. There are many guides about [writing](https://github.com/nelsonic/learn-mocha) [JavaScript](http://alistapart.com/article/writing-testable-javascript) [unit](http://www.smashingmagazine.com/2012/06/27/introduction-to-javascript-unit-testing/) [tests](http://www.htmlgoodies.com/beyond/javascript/testing-javascript-using-the-jasmine-framework.html) that serve as great introductions. At CFPB we like to separate our JavaScript into discrete [npm modules](https://github.com/cfpb/front-end/blob/master/npm.md). This makes testing easier because each module includes self-contained tests.
```
Comment (anselmbradford) [inline testing.md]: Maybe in place of `npm modules` say `CommonJS modules` to cover front-end/backend?

File: testing.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+- [Tape](https://www.npmjs.com/package/tape) is a [TAP](http://en.wikipedia.org/wiki/Test_Anything_Protocol)-producing test harness for Node with great asynchronous support. Check out the [Testling tape guide](https://ci.testling.com/guide/tape).
+
+Check out 18f's excellent [testing cookbook](http://18f.github.io/testing-cookbook) for more best practices.
+
+## Code coverage
+
+The goal is to have all of your code covered by unit tests. Tools known as code coverage reporters will tell you what percentage of your codebase is covered by tests. Owning A Home uses [grunt-mocha-istanbul](https://github.com/pocesar/grunt-mocha-istanbul) to [integrate istanbul with their Mocha tests](https://github.com/cfpb/owning-a-home/blob/3192922393461295edf9803eec677512704b75dc/Gruntfile.js#L388-L401). When OAH tests are run using `grunt test`, [Istanbul](https://gotwarlost.github.io/istanbul/) prints the coverage % to the command line and generates an interactive HTML code coverage report in `test/coverage`.
```
Comment (anselmbradford) [inline testing.md]: Put OAH in parentheses when it's introduced `Owning A Home (OAH)`

**Comment by ascott1** [comment]
> lol, don't diss CoffeeScript. Much of ES6 has been influenced by it. 

That's a fair point @sebworks!

**Comment by ascott1** [comment]
Going to merge after talking with @imuchnik. She's going to be updating this further once merged.

**Pull request merged**